rust/lance/src/dataset/write/update.rs

```diff
@@ -0,0 +1,566 @@
+///     .update_where("region_id = 10")
+///     .set("region_name", "New York")
+///     .build?
+///     .execute()
+///     .await?;
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct UpdateBuilder {
+    /// The dataset snapshot to update.
+    dataset: Arc<Dataset>,
+    /// The condition to apply to find matching rows to update. If None, all rows are updated.
+    condition: Option<Expr>,
+    /// The updates to apply to matching rows.
+    updates: HashMap<String, Expr>,
+}
+
+impl UpdateBuilder {
+    pub fn new(dataset: Arc<Dataset>) -> Self {
+        Self {
+            dataset,
+            condition: None,
+            updates: HashMap::new(),
+        }
+    }
+
+    pub fn update_where(mut self, filter: &str) -> Result<Self> {
+        let planner = Planner::new(Arc::new(self.dataset.schema().into()));
+        let expr = planner
+            .parse_filter(filter)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+        self.condition = Some(
+            planner
+                .optimize_expr(expr)
+                .map_err(box_error)
+                .context(InvalidInputSnafu)?,
+        );
+        Ok(self)
+    }
+
+    pub fn set(mut self, column: impl AsRef<str>, value: &str) -> Result<Self> {
+        let field = self
+            .dataset
+            .schema()
+            .field(column.as_ref())
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    format!(
+                        "Column '{}' does not exist in dataset schema: {:?}",
+                        column.as_ref(),
+                        self.dataset.schema()
+                    ),
+                    location!(),
+                )
+            })?;
+
+        // TODO: support nested column references. This is mostly blocked on the
+        // ability to insert them into the RecordBatch properly.
+        if column.as_ref().contains('.') {
+            return Err(Error::NotSupported {
+                source: format!(
+                    "Nested column references are not yet supported. Referenced: {}",
+                    column.as_ref(),
+                )
+                .into(),
+                location: location!(),
+            });
+        }
+
+        let schema: Arc<ArrowSchema> = Arc::new(self.dataset.schema().into());
+        let planner = Planner::new(schema.clone());
+        let mut expr = planner
+            .parse_expr(value)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+
+        // Cast expression to the column's data type if necessary.
+        let dest_type = field.data_type();
+        let df_schema = DFSchema::try_from(schema.as_ref().clone())?;
+        let src_type = expr
+            .get_type(&df_schema)
+            .map_err(box_error)
+            .context(InvalidInputSnafu)?;
+        if dest_type != src_type {
+            // TODO: remove this once DataFusion supports casting List to FSL
```

Comment:
Won't we still need `safe_coerce_scalar`?  Or does this TODO only refer to the `Expr::Literal(value @ ScalarValue::List` branch? (if so, maybe move the TODO above the branch)